### PR TITLE
Added additional OpenStack Authentication Values 

### PR DESCRIPTION
**Pull request by afritzler**
**What this PR does / why we need it**:
Added additional OpenStack Authentication Values 

**Which issue(s) this PR fixes**:
This PR is part of https://github.com/gardener/gardener-extensions/issues/154

**Release note**:
```noteworthyuser
Added support for additional OpenStack authentication values: `TenantID`, `DomainID`, `UserDomainName`, `UserDomainID`. Check out the example manifests.
```

### CODE REVIEW

File: controllers/provider-openstack/charts/internal/openstack-infra/templates/main.tf [hunk 1]
```diff
@@ -1,7 +1,23 @@
...
+  domain_name = "{{ .Values.openstack.domainName }}"
+  {{- end}}
+  {{- if .Values.openstack.domainID }}
+  domain_id   = "{{ .Values.openstack.domainID }}"
+  {{- end}}
+  {{- if .Values.openstack.tenantName }}
+  tenant_name = "{{ .Values.openstack.tenantName }}"
+  {{- end}}
+  {{- if .Values.openstack.tenantID }}
+  tenant_name = "{{ .Values.openstack.tenantID }}"
```
Comment (rfranzke) [inline controllers/provider-openstack/charts/internal/openstack-infra/templates/main.tf]:
```suggestion
  tenant_id = "{{ .Values.openstack.tenantID }}"
```

File: controllers/provider-openstack/charts/internal/openstack-infra/templates/main.tf [hunk 2]
```diff
@@ -1,7 +1,23 @@
...
+  {{- end}}
+  {{- if .Values.openstack.userDomainName }}
+  tenant_name = "{{ .Values.openstack.userDomainName }}"
```
Comment (rfranzke) [inline controllers/provider-openstack/charts/internal/openstack-infra/templates/main.tf]:
```suggestion
  user_domain_name = "{{ .Values.openstack.userDomainName }}"
```

File: controllers/provider-openstack/charts/internal/openstack-infra/templates/main.tf [hunk 3]
```diff
@@ -1,7 +1,23 @@
...
+  {{- end}}
+  {{- if .Values.openstack.userDomainID }}
+  tenant_name = "{{ .Values.openstack.userDomainID }}"
```
Comment (rfranzke) [inline controllers/provider-openstack/charts/internal/openstack-infra/templates/main.tf]:
```suggestion
  user_domain_id = "{{ .Values.openstack.userDomainID }}"
```

**Comment by rfranzke** [comment]
Can you check why the CI is failing?

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go [hunk 4]
```diff
@@ -240,6 +246,16 @@ func getConfigChartValues(
 		"requestTimeout":    cluster.CloudProfile.Spec.OpenStack.RequestTimeout,
 	}
 
+	// Only the 1.15 out of tree provider for OpenStack support the OS_DOMAIN_NAME/ID
+	// https://github.com/kubernetes/cloud-provider-openstack/pull/733
+	ok, err := gutils.CompareVersions(cluster.Shoot.Spec.Kubernetes.Version, ">=", "1.15")
```
Comment (rfranzke) [inline controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: We cannot merge this PR yet because there is no release of the OpenStack CCM, right?

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go [hunk 5]
```diff
@@ -240,6 +246,16 @@ func getConfigChartValues(
...
+	if ok {
+		values = map[string]interface{}{
```
Comment (rfranzke) [inline controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: Why are you overwriting the complete `values` map?

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go [hunk 6]
```diff
@@ -229,9 +231,13 @@ func getConfigChartValues(
 	values := map[string]interface{}{
 		"kubernetesVersion": cluster.Shoot.Spec.Kubernetes.Version,
 		"domainName":        c.DomainName,
+		"domainID":          c.DomainID,
 		"tenantName":        c.TenantName,
+		"tenantID":          c.TenantID,
 		"username":          c.Username,
 		"password":          c.Password,
+		"userDomainID":      c.UserDomainID,
```
Comment (rfranzke) [inline controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: Why are these added here? According to below comparison they are only valid for a certain Kubernetes version and should only be added if the version constraint matches?

File: controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go [hunk 7]
```diff
@@ -138,6 +138,15 @@ func (e *ensurer) getBackupRestoreContainer(name string, cluster *extensionscont
 						},
 					},
 				},
+				{
+					Name: "OS_DOMAIN_ID",
+					ValueFrom: &corev1.EnvVarSource{
+						SecretKeyRef: &corev1.SecretKeySelector{
+							Key:                  openstack.DomainID,
```
Comment (rfranzke) [inline controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go]: OK, so adding these env variables always doesn't harm, right?
However, what happens if this key does not exist in the referenced secret? Will the pod come up anyways and then the value of the env variable will be `""`?

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go (hunk 4 above)
Comment (afritzler) [reply controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: Yes I would wait. The merge should happen tomorrow.

File: controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go (hunk 7 above)
Comment (afritzler) [reply controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go]: The values are non pointer strings so yes, they will become "". Setting those envs doesn't harm, as the gophercloud authenticator reads and evaluates the environment variables correctly. https://github.com/gophercloud/gophercloud/blob/master/openstack/auth_env.go#L36

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go (hunk 5 above)
Comment (afritzler) [reply controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: That was a mistake. Wanted to add those keys.

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go (hunk 4 above)
Comment (rfranzke) [reply controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: Even if it's merged we have to wait for a new release. Do you know how often they are releasing?

File: controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go (hunk 7 above)
Comment (rfranzke) [reply controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go]: Hm, actually, the pod won't come up if the referenced keys don't exist in the secret. Just ran into this problem (cc @swapnilgm).

Comment (swapnilgm) [reply controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go]: Yes. It shows "ContainerConfigError".

Comment (afritzler) [reply controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go]: Ok, then I need to make this conditional.

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go (hunk 6 above)
Comment (afritzler) [reply controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: No, the domain and tenant ID is available for old cloud provider versions as well.

File: controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go (hunk 7 above)
Comment (afritzler) [reply controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go]: So this part need to be completely refactored. I added here secrets file and dir handling to the etcd-backup pod. That way we can just dump the correct secrets into the pod without having to fiddle around with the pod spec validation. The OpenStack authentication mechanics are then handling the authentication correctly. https://github.com/gardener/gardener-extensions/pull/287

Comment (afritzler) [reply controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go]: @swapnilgm I just updated the PR with the correct handling of `cloud.yaml` in the etcd-backup pod spec.

File: controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go [hunk 8]
```diff
@@ -110,57 +110,21 @@ func (e *ensurer) getBackupRestoreContainer(name string, cluster *extensionscont
 			provider = openstack.StorageProviderName
 			env = []corev1.EnvVar{
 				{
-					Name: "STORAGE_CONTAINER",
+					Name: "OS_CLIENT_CONFIG_FILE",
```
Comment (swapnilgm) [inline controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go]: You will still have to pass bucketName via STORAGE_CONTAINER env.

Comment (afritzler) [reply controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go]: Good point.

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go (hunk 4 above)
Comment (afritzler) [reply controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: So they are releasing not that often. However we will push for a release once all PRs are merged (still pending) on the cloud controller side. However we could move forward with this even if the CCM isn't available yet. You are just not allowed to set the user-domain-name. :)

Comment (rfranzke) [reply controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: What would happen if an end-user still specifies the user-domain-name field? Would the current CCM crash, or would simply happen "nothing"?

**Comment by rfranzke** [comment]
@swapnilgm PTAL: https://github.com/gardener/gardener-extensions/pull/287/files#diff-97489784c4edd1edc4cc864e568ed184R50

File: hack/test.sh [hunk 9]
```diff
@@ -24,4 +24,8 @@ SKIP_TESTS=$(echo controllers/provider-{alicloud,aws,azure,gcp,openstack,packet}
 
 header_text "Test"
 
+if [ $# -gt 0 ]; then
```
Comment (rfranzke) [inline hack/test.sh]: What is this thing doing? Why do we need it?

File: controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go [hunk 10]
```diff
@@ -121,46 +121,19 @@ func (e *ensurer) getBackupRestoreContainer(name string, cluster *extensionscont
 					},
 				},
 				{
-					Name: "OS_AUTH_URL",
+					Name: "OS_CLIENT_CONFIG_FILE",
 					ValueFrom: &corev1.EnvVarSource{
```
Comment (swapnilgm) [inline controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go]:
```suggestion
					Value: path.Join("~/.config/openstack/", openstack.CloudYAML),
```

File: controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go [hunk 11]
```diff
@@ -121,46 +121,19 @@ func (e *ensurer) getBackupRestoreContainer(name string, cluster *extensionscont
...
-				},
-				{
-					Name: "OS_TENANT_NAME",
-					ValueFrom: &corev1.EnvVarSource{
-						SecretKeyRef: &corev1.SecretKeySelector{
-							Key:                  openstack.TenantName,
+							Key:                  openstack.CloudYAMLKey,
 							LocalObjectReference: corev1.LocalObjectReference{Name: openstack.BackupSecretName},
 						},
 					},
```
Comment (swapnilgm) [inline controllers/provider-openstack/pkg/webhook/controlplanebackup/ensurer.go]:
```suggestion
					},
			}
			volumeMounts = []corev1.VolumeMount{
				{
					Name:      openstack.BackupSecretName,
					MountPath: "~/.config/openstack",
				},
			}
```

File: controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go (hunk 4 above)
Comment (afritzler) [reply controllers/provider-openstack/pkg/controller/controlplane/valuesprovider.go]: The CCM will crashloop since the config file marshalling can't handle unknown properties. :(

**Comment by swapnilgm** [comment]
Hey sorry for late comment, but I was not sure so didn't pitch it earlier. But thinking more about i am more into favour of replacing [this](https://github.com/gardener/gardener/blob/master/hack/templates/resources/70-deprecated-secret-cloudprovider.yaml.tpl#L76) and [this](https://github.com/gardener/gardener/blob/master/hack/templates/resources/45-deprecated-secret-seed-backup.yaml.tpl#L73) with 
```Yaml
  clouds.yaml: base64(clouds.yaml)
``` 
- The `clouds.yaml` is standard openstack [configuration file](https://docs.openstack.org/os-client-config/latest/user/configuration.html) format. 
-  With current approach we will have to parse the different fields in auth info and pass it at different place. There will be too many `if else` block repeated at many places. 
- I'm not sure does terraform, MCM support it yet, if not we can add support there.
- This is on similar thought of google `serviceaccount.json`.
- For aws or azure or alicloud there is not much complexity so its fine. (like having different combination of auth fields).

cc: @rfranzke @afritzler

**Comment by kayrus** [comment]
> clouds.yaml: base64(clouds.yaml)

I'd rather use plain text.

> I'm not sure does terraformmer, MCM (cc @prashanth26 ) support it yet, if not we can add support there.
1.15.x does, but not all the auth options. All the options are supported in master branch, and will be supported in 1.16.x

**Comment by afritzler** [comment]
On hold as we first need to add the OpenStack CSI support to the OpenStack provider.

**Comment by rfranzke** [comment]
OK, please reopen the PR once it's ready to be reviewed and merged again. Thank you!
